### DISCUSSION TITLE
Modified Module Database Migrations with Oqtane v2.1+?

### DISCUSSION THREAD

**Discussion by ghhv** [oqtane/oqtane.framework#2059] [Q&A]
Is there any updated documentation or clues on how to apply module database migrations with Oqtane v2.1+ (please!)?

I saw Shaun's comments to see what's done in the Oqtane.Blog module. It seems just manually adding [the migration code](https://github.com/oqtane/oqtane.blogs/blob/master/Server/Migrations/01000401_AddBlogSummary.cs). 

I have created a default scaffolded new module and added one new field. I have manually added the migration code but how is it successfully triggered?

With the migration code added, my existing modules now fail with this error yet the database has not been updated. 
```
System.ArgumentNullException: Value cannot be null. (Parameter 'source')
   at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Linq.OrderedEnumerable`2..ctor(IEnumerable`1 source, Func`2 keySelector, IComparer`1 comparer, Boolean descending, OrderedEnumerable`1 parent)
   at System.Linq.Enumerable.OrderBy[TSource,TKey](IEnumerable`1 source, Func`2 keySelector)
   at Eqp.EqpMaster0903.Services.EqpMaster0903Service.GetEqpMaster0903sAsync(Int32 ModuleId) in C:\Data\git\Oqtane\OqtaneV3.0.3-March9\Eqp.EqpMaster0903\Client\Services\EqpMaster0903Service.cs:line 21
   at Eqp.EqpMaster0903.Index.OnInitializedAsync() in C:\Data\git\Oqtane\OqtaneV3.0.3-March9\Eqp.EqpMaster0903\Client\Modules\Eqp.EqpMaster0903\Index.razor:line 53
```
(this code is trying to return the existing list of data items in the module - there is one item)

I see that there is code to do a migration as part of the module install in Oqtane.Framework and a new site/tenant install. 

So I tried adding a new site with new tenant since it should run the full migration on a fresh database but it gives me an error that the existing module database already exists.. 
![image](https://user-images.githubusercontent.com/13160762/158047005-b7417cf7-e737-425f-b7ee-2c94d3b2d306.png)

I also saw these issues from Oqtane 2.x
[Unable to Initialize Module Database from Module Manager in C# code](https://github.com/oqtane/oqtane.framework/issues/666)
[DB Migrations for Custom Modules](https://github.com/oqtane/oqtane.framework/discussions/826)

Given the code seems to have changed, I wonder if "a proper solution" has since been implemented?

**Comment by sbwalker** [comment]
@ghhv in order to provide an appropriate answer, can you please indicate if you are familiar with Entity Framework Migrations? In the 2.1 release Oqtane was enhanced to support multiple database platforms - and this required a transition to using Migrations. The core framework uses this approach and there is support for modules as well. It is important to note that it is not required for modules. If you are only planning on using a specific database platform such as SQL Server in your environment then you are still able to use the script-based approach which was utilized in the 1.x release.

**Reply by ghhv** [reply]
@sbwalker Yes, very familiar with Entity Framework Migrations and have always defaulted to code-first development. The `Add-Migration` and `Update-Database` were not options here as you've mentioned previously for the multi-tenant issue. So in my test code, I have added the extra migration code to add an extra database field and the model has been changed. 
So how does this get applied if we cannot run  `Update-Database`? 
The `.Migrate(...)` code only seems to run on install?
I plan to use Postgres and a graph database going forward but Postgres is broken as per the other [date/time issue](https://github.com/oqtane/oqtane.framework/issues/2038) - I will update that outstanding issue soon.

**Reply by sbwalker** [reply]
@ghhv Oqtane uses the code-first approach. If you create a module using the Create Module option in the framework  you will see the following pattern in the scaffolded Server project:

```
/Migrations
  01000000_InitializeModule.cs
  /EntityBuilders
    ModuleEntityBuilder.cs
```

The ModuleEntityBuilder.cs defines the EntityBuilder for the module and any code-first migration logic for the initial creation of the module's database tables, keys, indexes, etc... The 01000000_InitializeModule.cs contains the logic to actually execute the initial migrations in ModuleEntityBuilder.cs (ie. it contains the Up() method). Note that there can be many migration classes defined in the /Migrations folder - each of them based on version numbers defined in an attribute within the class ie.

[Migration("Module.01.00.00.00")]

The version number should correspond to a "release" of a module. So once the module has been initially installed onto a system and needs to be upgraded to include another column you need to create another "version" of the module which includes a migration class ie.

01000100_AddColumn.cs

and the attribute within that class would need to be:

[Migration("Module.01.00.01.00")]

The migration needs to implement the Up() method:

        protected override void Up(MigrationBuilder migrationBuilder)
        {
            var moduleEntityBuilder = new ModuleEntityBuilder(migrationBuilder, ActiveDatabase);
            moduleEntityBuilder.AddStringColumn("YourColumn", 50, true);
        }

During framework runtime startup it iterates through all of the databases which are part of the installation (ie. in a multi-tenant installation there may be many databases) and it executes any migrations that are required in each. It determines this based on the version of the module currently installed (ie. the ModuleDefinitions table contains a column which indicates the current version of a module which is installed). So simply adding a new migration class to your module is not enough for the framework to know it needs to run it. You also need to set the appropriate version properties in your IModule interface class ie.

```
        public ModuleDefinition ModuleDefinition => new ModuleDefinition
        {
            Name = "Module",
            Description = "Module",
            Version = "1.0.1",
            ServerManagerType = "Oqtane.Module.Manager.ModuleManager, Oqtane.Module.Server.Oqtane",
            ReleaseVersions = "1.0.0,1.0.1",
            Dependencies = "Oqtane.Module.Shared.Oqtane",
            SettingsType = "Oqtane.Module.Settings, Oqtane.Module.Client.Oqtane"
      }
```
Version and ReleaseVersions are both critical so that the framework is able to determine the various official versions releases of the module which have occurred over time - so that it can execute all upgrade logic in the appropriate order.

If you need to diagnose issues in your migration code you can look at the Server /Content/Log/error.log file which is a log sink during startup. You can also look at the standard __EFMigrationsHistory table in each database to determine if your migration has been executed.

**Reply by werddomain** [reply]
this info must be added to the documentation 😅

**Reply by mkashif14** [reply]
I agree, this is valuable info and must be added to the documentation.
